rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+should still work, but it should just be an alias for
+`include_sensitive_results` and should be marked as deprecated in or altogether
+excluded from the API documentation.
+
+##### Implementation details
+
+This will likely be implemented as a secondary index of results that do not
+include the sensitive terms, as explored in
+https://github.com/WordPress/openverse-api/pull/1108.
+
+#### Designation of results with sensitive terms
+
+Results that include sensitive terms in their textual content are so denoted in
+the API results in a new result field named `sensitivity`. Sensitivity should be
+an array that should be empty for results where `mature = false` and the textual
+content does not contain any sensitive terms. For results marked as `mature`,
+the array should include a `"mature"` string. For results with sensitive terms
+in their textual content, the array should include a `"sensitive_text"` string.
+At this time we will not be denoting _what_ sensitive terms are present nor will
+we be developing any categories of terms. This approach leaves open the
+potential for that in the future, however, by allowing the array to include
+specific tokens for particular categories of sensitivity, whether derived from a
+categorisation of the sensitive terms list, or from an image classification API.
+
+Note that this designation needs to happen both in the search endpoint and in
+the single result endpoint. Each endpoint may need a different approach to
+achieve this. In particular, the approach taken will depend on whether
+Elasticsearch is used to determine whether a result needs to be designated as
+having sensitive textual content or if this is done in pure Python.
+
+##### Implementation details
+
+There are two broad approaches that can be taken for this. I am actively
+consulting with people more familiar with Elasticsearch for the best way to do
+this, but the broad strokes of this are that we will either:
+
+- Loop over results in Python and using a Regex to determine if textual content
+  on the result includes sensitive terms. The results of this will be cached in
+  Redis to ameliorate performance degradation over time.
+- Use a script field and multi-index search to determine in Elasticsearch and as
+  a hit property whether the result is included in the filtered index (and is
+  therefore "safe").
+
+There may also be other viable ways of performing this determination in
+Elasticsearch, but the detail remains. If the determination is made in Python,
+then it is easy to share the implementation between the search endpoint and the
+single result endpoint (and both can benefit from the Redis caching). If the
+determination is made in Elasticsearch and benefits from Elasticsearch's full
+text search, then we may need to change the single result endpoint to pull the
+result in question from Elasticsearch rather than Postgres, leveraging the
+`_source` property on hits. This could be a big, complex change, though and
```

Comment:
I am not sure we have access to the `_source` property on hits. @dhruvkb, could you confirm?

Reply:
Elasticsearch DSL does indeed include `_source` on the Hit data: https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/response/hit.py#L24-L27